File: simulator/effects/effect_tracker.py
import logging

from ..effects.aoe_square_effect import AoeSquareEffect
from ..effects.effect import EffectType
from ..effects.post_haste_lethargy_effect import PostHasteLethargyEffect
from ..effects.aoe_spheric_effect import AoeSphericEffect

logger = logging.getLogger("Encounterra")


class EffectTracker:
    """
    Manages all lasting effects.
    """
    def __init__(self):
        self.effects = []
        self.battle_map = None

    def add(self, effect):
        self.effects.append(effect)

    def remove(self, effect):
        effect.deactivate()
        self.effects.remove(effect)

    def start_of_turn_tick(self, combatant):
        """
        Manages all effects with a fixed duration measurable in rounds which end just before the beginning of combatant's turn.
        :param combatant: the initiating combatant
        :return:
        """
        effects = []
        for e in self.effects:
            if e.initiator is combatant:
                try:
                    keep = e.start_of_turn_tick()
                    if not keep:
                        e.deactivate()
                        continue  # Effect expired
                except AttributeError:  # This skips
                    pass
            effects.append(e)  # Effect persists
        self.effects = effects

    def start_of_turn(self, combatant):
        """
        Manages effects which take effect at the start of a combatant's turn.
        Also manages effects that can be saved against at the start of a combatant's turn.
        :param combatant: the affected combatant
        :return:
        """
        effects = []
        for e in self.effects:
            if e.is_affecting(combatant):
                if not e.start_of_turn_for_combatant(combatant):
                    e.deactivate_for_combatant(combatant)  # TODO: at the moment this does nothing (regeneration and digestion only)
                    continue  # Effect's been cancelled
            effects.append(e)  # Effect persists
        self.effects = effects

    def end_of_turn(self, combatant):
        effects = []
        for e in self.effects:
            if e.is_affecting(combatant):
                if not e.combatant_saved_at_end_of_turn(combatant):
                    if not e.deactivate_for_combatant(combatant):
                        continue  # Effect's been saved against or somehow ceased on all combatants -> can be removed
            effects.append(e)
        self.effects = effects

    def get_affecting_combatant(self, combatant):
        """
        Returns a set of all effects affecting a combatant
        :param combatant: the combatant
        :return:  set of all effects affecting the combatant
        """
        try:
            return {e for e in self.effects if e.is_affecting(combatant)}
        except AttributeError:
            print("FIXME")

    def is_affecting_combatant(self, combatant, effect_type):
        """
        Determines whether a combatant is affected by an effect of a certain type
        :param combatant:
        :param effect_type: EffectType enum
        :return: True if the combatant is affected, False otherwise
        """
        for e in self.effects:
            if e.get_effect_type() is effect_type and e.is_affecting(combatant):
                return True
        return False

    def get_aoe_effects(self):
        return [e for e in self.effects if isinstance(e, AoeSquareEffect) or isinstance(e, AoeSphericEffect)]

    def get_effects_by_initiator(self, initiator):
        return [e for e in self.effects if e.initiator is initiator]

    def combatant_died(self, combatant):
        for e in self.get_effects_by_initiator(combatant):
            e.deactivate()
        self.effects = [e for e in self.effects if e.initiator is not combatant]

    def create_post_haste_lethargy(self, initiator, combatant):
        self.effects.append(PostHasteLethargyEffect(initiator, combatant))

# TODO add function for wildshape replacement

    def remove_effect_from_combatant_by_type(self, combatant, effect_type):
        effects = []
        for e in self.effects:
            if e.is_affecting(combatant) and e.get_effect_type() is effect_type:
                if not e.deactivate_for_combatant(combatant):
                    continue
            effects.append(e)
        self.effects = effects

    def remove_effect_from_combatant(self, combatant, effect):
        if not effect.deactivate_for_combatant(combatant):
            self.effects.remove(effect)

    def is_combatant_hidden_from(self, combatant, target):
        """

        :param combatant: the hiding combatant
        :param target: the one hiding from
        :return: True if the combatant is hidden from target
        """
        for e in self.get_effects_by_initiator(combatant):
            if e.get_effect_type() is EffectType.HIDE and e.target is target:
                return True
        return False

    def is_affected_by_vow_of_enmity(self, initiator, target):
        for e in self.get_effects_by_initiator(initiator):
            if e.get_effect_type() is EffectType.VOW_OF_ENMITY and e.combatants[0] is target:
                return True
        return False

    def reset(self):
        for effect in self.effects:
            effect.deactivate()
        self.effects.clear()

File: c++/effects/effect_tracker.cpp
#include "effects/effect_tracker.hpp"
#include "effects/aoe_square_effect.hpp"
#include "effects/aoe_spheric_effect.hpp"
#include <algorithm>

namespace enc
{
  std::weak_ptr<Effect> EffectTracker::add(std::shared_ptr<Effect> effect)
  {
    // Avoid double-tracking the same effect. Concentration spells call setConcentrationEffect()
    // during activate(), which re-adds an effect the action resolver already inserted. A duplicate
    // entry would be processed twice each round and linger as a zombie after the original is removed.
    if(std::find(_effects.begin(), _effects.end(), effect) != _effects.end())
      {
        return std::weak_ptr<Effect>(effect);
      }
    _effects.push_back(effect);
    return std::weak_ptr<Effect>(effect);
  }

  void EffectTracker::remove(const std::shared_ptr<Effect> &effect)
  {
    effect->deactivate();
    _effects.erase(std::remove(_effects.begin(), _effects.end(), effect), _effects.end());
  }

  void EffectTracker::startOfTurnTick(Combatant *combatant)
  {
    // Iterate over a snapshot: deactivate() on a concentration effect calls breakConcentration(),
    // which removes from _effects and would invalidate a live iterator over _effects.
    std::vector<std::shared_ptr<Effect>> snapshot = _effects;
    std::vector<std::shared_ptr<Effect>> remainingEffects;

    for(const auto &effect : snapshot)
      {
        if(effect->getInitiator() == combatant)
          {
            bool keep = effect->startOfTurnTick();
            if(!keep)
              {
                effect->deactivate();
                continue; // Effect expired
              }
          }
        remainingEffects.push_back(effect); // Effect persists
      }
    _effects = std::move(remainingEffects);
  }

  void EffectTracker::startOfTurn(Combatant *combatant)
  {
    // Iterate over a snapshot: deactivateForCombatant() on a concentration effect calls
    // breakConcentration(), which removes from _effects and would invalidate a live iterator.
    std::vector<std::shared_ptr<Effect>> snapshot = _effects;
    std::vector<std::shared_ptr<Effect>> remainingEffects;

    for(const auto &effect : snapshot)
      {
        if(effect->isAffecting(combatant))
          {
            if(!effect->startOfTurnForCombatant(combatant))
              {
                effect->deactivateForCombatant(combatant);
                continue; // Effect's been cancelled
              }
          }
        remainingEffects.push_back(effect);
      }
    _effects = std::move(remainingEffects);
  }

  void EffectTracker::endOfTurn(Combatant *combatant)
  {
    // Iterate over a snapshot: a successful save calls deactivateForCombatant(), which for a
    // concentration spell calls breakConcentration() -> remove(), mutating _effects mid-loop.
    // Iterating _effects directly would invalidate the iterator and crash.
    std::vector<std::shared_ptr<Effect>> snapshot = _effects;
    std::vector<std::shared_ptr<Effect>> remainingEffects;

    for(const auto &effect : snapshot)
      {
        if(effect->isAffecting(combatant))
          {
            if(!effect->combatantSavedAtEndOfTurn(combatant))
              {
                if(!effect->deactivateForCombatant(combatant))
                  {
                    continue; // Effect's been saved against or ceased
                  }
              }
          }
        remainingEffects.push_back(effect);
      }
    _effects = std::move(remainingEffects);
  }

  std::unordered_set<std::shared_ptr<Effect>> EffectTracker::getAffectingCombatant(Combatant *combatant) const
  {
    std::unordered_set<std::shared_ptr<Effect>> affectingEffects;
    for(const auto &effect : _effects)
      {
        if(effect->isAffecting(combatant))
          {
            affectingEffects.insert(effect);
          }
      }
    return affectingEffects;
  }

  bool EffectTracker::isAffectingCombatant(Combatant *combatant, EffectType effectType) const
  {
    for(const auto &effect : _effects)
      {
        if(effect->getEffectType() == effectType && effect->isAffecting(combatant))
          {
            return true;
          }
      }
    return false;
  }

  std::vector<std::shared_ptr<AoeEffect>> EffectTracker::getAoeEffects() const
  {
    std::vector<std::shared_ptr<AoeEffect>> aoeEffects;
    for(const auto &effect : _effects)
      {
        if(auto aoeEffect = std::dynamic_pointer_cast<AoeEffect>(effect))
          {
            aoeEffects.push_back(aoeEffect);
          }
      }
    return aoeEffects;
  }

  std::vector<std::shared_ptr<Effect>> EffectTracker::getEffectsByInitiator(Combatant *initiator) const
  {
    std::vector<std::shared_ptr<Effect>> initiatorEffects;
    for(const auto &effect : _effects)
      {
        if(effect->getInitiator() == initiator)
          {
            initiatorEffects.push_back(effect);
          }
      }
    return initiatorEffects;
  }

  std::vector<std::shared_ptr<Effect>> EffectTracker::getEffectsByType(EffectType effectType) const
  {
    std::vector<std::shared_ptr<Effect>> matching;
    for(const auto &effect : _effects)
      {
        if(effect->getEffectType() == effectType)
          {
            matching.push_back(effect);
          }
      }
    return matching;
  }

  void EffectTracker::combatantDied(Combatant *combatant)
  {
    // Work from a snapshot because deactivating concentration effects can recursively mutate _effects.
    std::vector<std::shared_ptr<Effect>> snapshot = _effects;
    std::vector<std::shared_ptr<Effect>> remainingEffects;

    for(const auto &effect : snapshot)
      {
        if(effect->getInitiator() == combatant)
          {
            effect->deactivate();
            continue;
          }
        else if(effect->isAffecting(combatant))
          {
            if(!effect->deactivateForCombatant(combatant))
              {
                continue;
              }
          }

        remainingEffects.push_back(effect);
      }
    _effects = std::move(remainingEffects);
  }

  // void createPostHasteLethargy(Combatant *initiator, Combatant *combatant)
  // {
  //   _effects.push_back(std::make_shared<PostHasteLethargyEffect>(initiator, combatant));
  // }

  void EffectTracker::removeEffectFromCombatantByType(Combatant *combatant, EffectType effectType)
  {
    std::vector<std::shared_ptr<Effect>> remainingEffects;
    for(const auto &effect : _effects)
      {
        if(effect->isAffecting(combatant) && effect->getEffectType() == effectType)
          {
            if(!effect->deactivateForCombatant(combatant))
              {
                continue;
              }
          }
        remainingEffects.push_back(effect);
      }
    _effects = std::move(remainingEffects);
  }

  void EffectTracker::removeEffectFromCombatant(Combatant *combatant, const std::shared_ptr<Effect>& effect)
  {
    if(!effect->deactivateForCombatant(combatant))
      {
        _effects.erase(std::remove(_effects.begin(), _effects.end(), effect), _effects.end());
      }
  }

  bool EffectTracker::isCombatantHiddenFrom(Combatant *combatant, Combatant *target) const
  {
    for(const auto &effect : getEffectsByInitiator(combatant))
      {
        if(effect->getEffectType() == EffectType::HIDE && effect->getTarget() == target)
          {
            return true;
          }
      }
    return false;
  }

//   bool EffectTracker::isAffectedByVowOfEnmity(Combatant *initiator, Combatant *target) const
//   {
//     for(const auto &effect : getEffectsByInitiator(initiator))
//       {
//         if(effect->getEffectType() == EffectType::VOW_OF_ENMITY && !effect->getCombatants().empty() && effect->getCombatants()[0] == target)
//           {
//             return true;
//           }
//       }
//     return false;
//   }

  void EffectTracker::reset()
  {
    // Iterate over a snapshot: deactivate() on a concentration effect calls breakConcentration(),
    // which removes from _effects and would invalidate a live iterator over _effects (mirrors
    // startOfTurnTick). The final clear() guarantees the tracker ends empty for the next iteration.
    std::vector<std::shared_ptr<Effect>> snapshot = _effects;
    for(const auto &effect : snapshot)
      {
        effect->deactivate();
      }
    _effects.clear();
  }
}
